### PR TITLE
Backports v0.21.0 (3)

### PR DESCRIPTION
**Pull request by haampie**
- [x] #38803
- [x] #40934 (currently it's backports/v0.21.0)
- [x] #40941 
- [x] #40959
- [x] #40998
- [x] #40997 
- [x] #41002 
- [x] #41003
- [x] Set version to 0.21.0
- [x] Update `CHANGELOG.md` for 0.21.0

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,249 @@
+# v0.21.0 (2023-11-10)
+
+`v0.21.0` is a major feature release.
+
+## Features in this release
+
+1. **Better error messages with Condition chaining**
```
Comment (haampie) [inline CHANGELOG.md]: upper case intended?

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,249 @@
...
+          required because hdf5 ^cmake@3.0.1 requested from CLI
+   4. Cannot satisfy 'cmake@3.12:' and 'cmake@3.0.1
+        required because hdf5 depends on cmake@3.12:
+          required because hdf5 ^cmake@3.0.1 requested from CLI
+        required because hdf5 ^cmake@3.0.1 requested from CLI
+```
+
+More details in #40173.
+
+2. **OCI buildcaches**
```
Comment (haampie) [inline CHANGELOG.md]: build cache or buildcache?

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,249 @@
...
+cache:
+
+```console
+$ spack mirror add my_registry oci://user/image # Dockerhub
+$ spack mirror add my_registry oci://ghcr.io/haampie/spack-test # GHCR
+$ spack mirror set --push --oci-username ... --oci-password ... my_registry  # set login credentials
+$ spack buildcache push my_registry [specs...]
+```
+
+And you can add optinally add a base image to get *runnable* images:
```
Comment (haampie) [inline CHANGELOG.md]:
```suggestion
And you can add optionally add a base image to get *runnable* images:
```

**Review by haampie — COMMENTED** [review]
not sure if these review comments appear as i write them

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,3 +1,249 @@
...
+```
+
+and have everything that depends on `meson` rebuilt the next time you run `spack
+concretize`. In a future Spack version, we'll handle all of this in a single command,
+but for now you can use this to drop bits of your lockfile and resolve your dependencies
+again. More in #38803.
+
+7. **New `spack info` UI**
+
+The venerable `spack info` commannd was looking shabby compared to the rest of Spack's
```
Comment (haampie) [inline CHANGELOG.md]:
```suggestion
The venerable `spack info` command was looking shabby compared to the rest of Spack's
```

**Review by haampie — COMMENTED** [review]
.

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,3 +1,252 @@
...
+
+* archspec: update to v0.2.2, support for Sapphire Rapids, Power10, Neoverse V2 (#40917)
+* Propagate variants across nodes that don't have that variant (#38512)
+* Implement fish completion (#29549)
+* Can now distinguish between source/binary mirror; don't ping mirror.spack.io as much (#34523)
+* Improve status reporting on install (add [n/total] display) (#37903)
+
+## Windows
+
+This release has the best windows support of any Spack release yet, with numerous
```
Comment (haampie) [inline CHANGELOG.md]:
```suggestion
This release has the best Windows support of any Spack release yet, with numerous
```

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,3 +1,252 @@
...
+* Windows: executable/path handling (#37762)
+* Windows build systems: use ninja and enable tests (#33589)
+* Windows testing (#36970, #36972, #36973, #36840, #36977, #36792, #36834, #34696, #36971)
+* Windows PowerShell support (#39118, #37951)
+* Windows symlinking and libraries (#39933, #38599, #34701, #38578, #34701)
+
+## Notable refactors
+* User-specified flags take precedence over others in Spack compiler wrappers (#37376)
+* Improve setup of build, run, and test environments (#35737, #40916)
+* AutotoolsPackage / MakefilePackage: add gmake build dependency (#40380)
```
Comment (haampie) [inline CHANGELOG.md]:
```suggestion
* `make` is no longer a required system dependency of Spack (#40380)
```

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,3 +1,252 @@
...
+* Support Python 3.12 (#40404, #40155, #40153)
+* docs: Replace package list with packages.spack.io (#40251)
+* Drop Python 2 constructs in Spack (#38720, #38718, #38703)
+* Remove `LazyReference` from code (#38944)
```
Comment (haampie) [inline CHANGELOG.md]:
```suggestion
```
dunno if that's super relevant?

File: CHANGELOG.md [hunk 8]
```diff
@@ -1,3 +1,252 @@
...
+* Module file generation is disabled by default; you'll need to enable it to use it (#37258)
+* Remove deprecated "extra_instructions" option for containers (#40365)
+* Stand-alone test feature deprecation postponed to v0.22 (#40600)
+* buildcache push: make `--allow-root` the default and deprecate the option (#38878)
+
+## Notable Bugfixes
+* Bugfix: propagation of multivalued variants (#39833)
+* Allow `/` in git versions (#39398)
+* Fetch & patch: actually acquire stage lock, and many more issues (#38903)
+* Environment/depfile: fix bug with Git hash versions (attempt #2) (#37560)
```
Comment (haampie) [inline CHANGELOG.md]:
```suggestion
* Environment/depfile: better escaping of targets with Git versions (#37560)
```

File: CHANGELOG.md [hunk 9]
```diff
@@ -1,3 +1,252 @@
...
+And you can add optinally add a base image to get *runnable* images:
+
+```console
+$ spack buildcache push --base-image ubuntu:23.04 my_registry python
+Pushed ... as [image]:python-3.11.2-65txfcpqbmpawclvtasuog4yzmxwaoia.spack
+
+$ docker run --rm -it [image]:python-3.11.2-65txfcpqbmpawclvtasuog4yzmxwaoia.spack
+```
+
+More details in #38358.
```
Comment (haampie) [inline CHANGELOG.md]:
```suggestion
This creates a container image from the Spack installations on the host system, without the
need to run `spack install` from a `Dockerfile` or `sif` file. It also addresses the inconvenience
of losing binaries of dependencies when `RUN spack install` fails inside `docker build`.

Further, the container image layers and build cache tarballs are the same files. This means that
`spack install` and `docker pull` use the exact same underlying binaries. If you previously used
`spack install` inside of `docker build`, this feature helps you save storage by a factor two.

More details in #38358.
```

**Review by haampie — COMMENTED** [review]
Some more explanation of oci stuff?

File: CHANGELOG.md (hunk 1 above)
Comment (tgamblin) [reply CHANGELOG.md]:
```suggestion
1. **Better error messages with condition chaining**
```

File: CHANGELOG.md (hunk 2 above)
Comment (tgamblin) [reply CHANGELOG.md]:
```suggestion
2. **OCI build caches**
```

File: CHANGELOG.md [hunk 10]
```diff
@@ -1,3 +1,280 @@
...
+cache:
+
+```console
+$ spack mirror add my_registry oci://user/image # Dockerhub
+$ spack mirror add my_registry oci://ghcr.io/haampie/spack-test # GHCR
+$ spack mirror set --push --oci-username ... --oci-password ... my_registry  # set login credentials
+$ spack buildcache push my_registry [specs...]
+```
+
+And you can add optionally add a base image to get *runnable* images:
```
Comment (haampie) [inline CHANGELOG.md]:
```suggestion
And you can optionally add a base image to get *runnable* images:
```

**Review by haampie — COMMENTED** [review]
one more thing

File: CHANGELOG.md [hunk 11]
```diff
@@ -1,3 +1,290 @@
...
+    dependencies. On top of that, you can safely use `hash_length: 0`, as this config
+    now only applies to the modules exposed to the user -- you don't have to worry about
+    file name clashes for hidden dependencies.
+
+   Note: for `tcl` this feature requires Modules 4.7 or higher
+
+11. **Updated container labeling**
+
+    Nightly Docker images from the `develop` branch will now be tagged as `:develop` and
+    `:nightly` The `:latest` tag no longer be associated with `:develop`, but with the
```
Comment (haampie) [inline CHANGELOG.md]:
```suggestion
    `:nightly`. The `:latest` tag is no longer associated with `:develop`, but with the
```

File: CHANGELOG.md [hunk 12]
```diff
@@ -1,3 +1,290 @@
...
+    latest stable release. Releases will be tagged with `:{major}`, `:{major}.{minor}`
+    and `:{major}.{minor}.{patch}` `ubuntu:18.04` has also been removed from the list of
```
Comment (haampie) [inline CHANGELOG.md]:
```suggestion
    and `:{major}.{minor}.{patch}`. `ubuntu:18.04` has also been removed from the list of
```

**Review by haampie — COMMENTED** [review]
Last typos

**Review by tgamblin — APPROVED** [review]

**Pull request merged**